[PA-210] Viewrecord

alex-demetriou opened the pull request:
### Adds a View endpoint for single records

Looking for ideas around table loading re: the relations between `tenagree`, `tenure` and `lookup`.
I wasn't able to get it working without: 
```
.Include(/*tenure*/)
.Include(/*lookup*/)
.ToList();
``` 
for some reason.

Commit:
AMended typo of subnetid to subnetids

makeyw commented on TenancyInformationApi.Tests/V1/Controllers/TenancyInformationControllerTests.cs:
```diff
@@ -0,0 +1,46 @@
+            response?.StatusCode.Should().Be(404);
+
+            // 123/1 has been added - should return OK200
+            _getByIdMock.Setup(x => x.Execute("123/1")).Returns(new TenancyInformationResponse());
```
I think your check on the controller (Line 45) sees if there's a tenancy ref and returns 404 if there's not. So you'll need to setup your returned object here with a tenancy ref as a minimum.

makeyw commented on TenancyInformationApi.Tests/V1/Controllers/TenancyInformationControllerTests.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        [Test]
+        public void ViewRecordReturnsResponseWithStatus()
```
I feel like all these checks should be separate tests perhaps?

makeyw reviewed (changes requested):
The other errors you're seeing are because the dummy database isn't being setup properly.
In the root of the repo there's a database folder with a Docker file and a .SQL file.

The SQL file will need to be setup with the schema for the tenagree table.

See: https://github.com/LBHackney-IT/housing-resident-information-api/blob/master/database/schema.sql as example (it has tenagree in there but you'll need your lookup tables as well).

The example tests should be deleted as well.

alex-demetriou commented:
> The other errors you're seeing are because the dummy database isn't being setup properly.
> In the root of the repo there's a database folder with a Docker file and a .SQL file.
> 
> The SQL file will need to be setup with the schema for the tenagree table.

That's really strange because when I connected to the local database earlier in the day, the tables were there and had everything they were supposed to.

Commit:
Add View endpoint (Single record)

humulla commented on TenancyInformationApi.Tests/V1/Controllers/TenancyInformationControllerTests.cs:
```diff
@@ -0,0 +1,55 @@
+            _getByIdMock.Setup(x => x.Execute("123/1")).Returns(new Fixture().Create<TenancyInformationResponse>());
+            // 123/1 has been added - should return OK200
+
+            var response = _classUnderTest.ViewRecord("123-1") as ObjectResult;
```
Would it not be `as OKobjectResult`

alex-demetriou replied:
`OkObjectResult` is more or less the same thing, but with a default `StatusCode` attribute.

I didn't want to inform the content returned by `_classUnderTest` from within the test itself.

Potentially we could have the following:
```suggestion
            var response = _classUnderTest.ViewRecord("123-1") as ObjectResult;
            response.Should().BeOfType<OkObjectResult>();
```

instead of the result.StatusCode comparison?

humulla replied:
No i think its okay to leave it to what is it, I didn't realise that was the difference!

humulla commented on TenancyInformationApi.Tests/V1/Gateways/TenancyGatewayTests.cs:
```diff
@@ -0,0 +1,67 @@
+
+    public static class UhEntityHelper
+    {
+        public static UhTenancyAgreement FromDomain<T>(T tenancy) where T : Tenancy =>
```
Should this not be in the Entity or Domain section (outside the gateway tests)

humulla commented on TenancyInformationApi.Tests/V1/UseCase/GetByIdUseCaseTests.cs:
```diff
@@ -1,23 +1,41 @@
+            var result = _classUnderTest.Execute(tenancyRef);
+
+            result.Should().NotBeNull();
+            result.Should().BeOfType<TenancyInformationResponse>();
```
I like this check!

humulla commented on TenancyInformationApi/Startup.cs:
```diff
@@ -141,9 +140,12 @@ public static void Configure(IApplicationBuilder app, IWebHostEnvironment env)
-            _apiVersions = api.ApiVersionDescriptions.ToList();
+            if (app != null)
+            {
+                //Get All ApiVersions,
```
Remove comment as no longer required

humulla commented on TenancyInformationApi/V1/Factories/TenancyResponseFactory.cs:
```diff
@@ -0,0 +1,29 @@
+            return new TenancyInformationResponse
+            {
+                TenancyReference = tenancy?.TenancyReference,
+                CommencementOfTenancyDate = tenancy?.CommencementOfTenancyDate.ToString(CultureInfo.CurrentCulture),
```
What if you try `ToString("O")`

alex-demetriou replied:
Good catch, I forgot to change this to match the swaggerdoc.
These dates should use `ToString("yyyy-MM-dd")` in that case.

```suggestion
                CommencementOfTenancyDate = tenancy?.CommencementOfTenancyDate.ToString("yyyy-MM-dd"),
```

humulla commented on TenancyInformationApi/V1/Infrastructure/UhTenancyAgreement.cs:
```diff
@@ -0,0 +1,29 @@
+    [Table("tenagree")]
+    public class UhTenancyAgreement
+    {
+        [StringLength(11)] [Column("tag_ref"), Key] public string TenancyAgreementReference { get; set; }
```
What if you format it as :
`[Column("tag_ref")]`
`[Key]`
`[StringLength(11)]`
`public string TenancyAgreementReference { get; set; }`
It may be easier to read

alex-demetriou replied:
Personally I didn't find that easier to read, but if there's a general consensus I'll change to it.

humulla commented:
Looks good just a couple of refactoring recomendations, let me know if you have any questions

alex-demetriou commented:
> Looks good just a couple of refactoring recomendations, let me know if you have any questions

:grin: Thanks for your feedback @humulla, let me know what you think of the proposed changes and I'll get them merged in.

elcorbs commented on TenancyInformationApi.Tests/V1/Controllers/TenancyInformationControllerTests.cs:
```diff
@@ -0,0 +1,55 @@
+            // 1231 is not a valid tag_ref - should return BadRequest400
+            var response = _classUnderTest.ViewRecord("1231") as ObjectResult;
+            response.Should().NotBeNull();
+            response?.StatusCode.Should().Be(400);
```
Might be good to also return a message here explaining the reason the 400 is thrown. Not too specific, just something like - Tag Ref isn't in a valid format

alex-demetriou replied:
@elcorbs should I write a test to make sure that message is present?

Currently I've got the controller responding `tag_ref is malformed or missing` for `400`s.

elcorbs replied:
yeah thats good, and yeah you can just add an assertion to this test

elcorbs commented on TenancyInformationApi.Tests/V1/Gateways/TenancyGatewayTests.cs:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public static class UhEntityHelper
```
This can be moved into it's own file - it's standard practice to only have one class per file

alex-demetriou replied:
That's me forgetting to remove unused code. Thanks for catching that :grin:

elcorbs commented on TenancyInformationApi/Startup.cs:
```diff
@@ -141,9 +140,12 @@ public static void Configure(IApplicationBuilder app, IWebHostEnvironment env)
-            //Get All ApiVersions,
-            var api = app.ApplicationServices.GetService<IApiVersionDescriptionProvider>();
-            _apiVersions = api.ApiVersionDescriptions.ToList();
+            if (app != null)
```
When would `app` be null?

alex-demetriou replied:
I don't know why, but when I built from the command line, I got an error for not having a null-check.
When I just ran the build in Rider, it didn't care after I removed it.

alex-demetriou replied:
Everything is building fine on CircleCi without the null-check, so away it goes!

elcorbs commented on TenancyInformationApi/V1/Boundary/Response/TenancyInformationResponse.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    public class TenancyInformationResponse
+    {
+        public string TenancyReference { get; set; }
```
Is this Tag Ref? I think the full name for Tag Ref is Tenancy Agreement Reference

elcorbs commented on TenancyInformationApi/V1/Controllers/TenancyInformationController.cs:
```diff
@@ -13,39 +14,52 @@ namespace TenancyInformationApi.V1.Controllers
+        // Two ideas:
+        [HttpGet, Route("{badTenRef}/{caught}")]
+        public IActionResult CatchSlashedReference(string badTenRef, string caught) =>
+            // BadRequest("Replace '/' with '-' in provided tag_ref.");
```
I think we should be returning BadRequest here instead of fixing the error. Accepting both formats feels like bad API design to me

elcorbs commented on TenancyInformationApi/V1/Gateways/TenancyGateway.cs:
```diff
@@ -0,0 +1,24 @@
+
+        public Tenancy GetById(string id)
+        {
+            return _uhContext
```
This is nice and succinct

elcorbs commented on TenancyInformationApi/V1/Infrastructure/UhContext.cs:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    [Table("lookup")]
+    public class UhAgreementType
```
Both of these tables should be in their own files

elcorbs commented on TenancyInformationApi/V1/Infrastructure/UhContext.cs:
```diff
@@ -0,0 +1,39 @@
+        public DbSet<UhTenancyAgreement> UhTenancyAgreements { get; set; }
+    }
+
+    [Table("lookup")]
```
If the table is in a schema which isn't `public`, this also needs to be specified here I think something like: `[Table("name"), SchemaName("dbo")]`

elcorbs commented on TenancyInformationApi/V1/Infrastructure/UhContext.cs:
```diff
@@ -0,0 +1,39 @@
+    {
+        public UhContext(DbContextOptions options) : base(options)
+        {
+            var tenancyAgreements = UhTenancyAgreements
+                .Include(t => t.UhAgreementType)
+                .Include(t => t.UhTenureType)
+                .ToList();
```
Is this bit of code used for anything?

elcorbs replied:
It is pulling everything from the tenancy agreement table, so is quite an expensive database operation to make

alex-demetriou replied:
The relation in EF wasn't working until I performed eager loading like this. I'm definitely interested in better solutions as, like you say, this is really expensive.

elcorbs commented on TenancyInformationApi/V1/Infrastructure/UhTenancyAgreement.cs:
```diff
@@ -0,0 +1,29 @@
+        [Column("cot")] public DateTime CommencementOfTenancy { get; set; }
+        [Column("eot")] public DateTime EndOfTenancy { get; set; }
+
+        public UhTenureType UhTenureType { get; set; }
+        public UhAgreementType UhAgreementType { get; set; }
```
For the foreign keys it is best practice to also have the Foreign Key Id explicitly defined here as well. I.e. have a field like  `public int UhTenureTypeId { get; set; }`

alex-demetriou replied:
There are properties for `UhTenureTypeId` and `UhAgreementTypeId` up top.
This is probably a good indication that they should be on adjacent lines.

elcorbs replied:
ah sorry I missed that!

elcorbs commented on TenancyInformationApi/V1/UseCase/GetTenancyByIdUseCase.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public TenancyInformationResponse Execute(string id)
+        {
+            var result = _gateway.GetById(id).ToResponse();
```
Could the result of `_gateway.GetById(id)` be null and hence throw an error when calling `.ToResponse()`?

alex-demetriou replied:
The assignments in `ToResponse()` all use null-conditional operators, so if there is no `Tenancy` returned by the Gateway, it will still return a new `TenancyInformationResponse` - just full of `null` members.

I've written a new test to make sure the Use Case still returns the object.

Commit:
Add suggestions/fixes

- Use `dbo` schema in place of default `public`
- Separate UH entities into individual files
- Use proper `yyyy-MM-dd` date format in response
- Rename `TenancyReference` to `TenancyAgreementReference`
- Respond `HTTP 400` for `tag_ref`s using forward slash
- Remove unused UhEntityHelper
- Add test around `tag_ref`s not in the database

Commit:
Add test assertions for correct HTTP 400/404 responses
